### ISSUE TITLE
False negative when passing invalid `lambda` to `Union` of higher-order functions

### ISSUE THREAD

**Issue by LeeeeT** [microsoft/pyright#7012]
When doing something with a `Union`, your code should be valid for every choice of the `Union`:

```python
type U = int | str

def f(u: U) -> None:
    u.lower()  # type checking error
```

Having first-order functions as the `Union` choices and passing a wrong argument to them also gives an error:

```python
from collections.abc import Callable

type Takes[T] = Callable[[T], object]

type U = Takes[int] | Takes[str]

def f(u: U) -> None:
    u(42)  # type checking error
```

But passing a **`lambda`** of a wrong type to a **`Union`** of **higher-order functions** does **not** produce an error:

```python
from collections.abc import Callable

type Takes[T] = Callable[[T], object]

type U = Takes[Takes[int]] | Takes[Takes[str]]

def f(u: U) -> None:
    u(lambda v: v.lower())  # no error :(
```

Mypy shows an error for the above code (substituting PEP 695 type aliases with the old ones, since they're not yet supported by mypy), and I expect pyright to do the same.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.348, which was just published. It will also be included in a future release of pylance.
